dataclients/kubernetes/kube.go

```diff
@@ -79,37 +92,130 @@ type Options struct {
 
 // Client is a Skipper DataClient implementation used to create routes based on Kubernetes Ingress settings.
 type Client struct {
+	httpClient         *http.Client
 	apiURL             string
 	provideHealthcheck bool
+	token              string
 	current            map[string]*eskip.Route
 }
 
 var nonWord = regexp.MustCompile("\\W")
 
 var (
-	errServiceNotFound     = errors.New("service not found")
-	errServicePortNotFound = errors.New("service port not found")
+	errServiceNotFound      = errors.New("service not found")
+	errServicePortNotFound  = errors.New("service port not found")
+	errAPIServerURLNotFound = errors.New("kubernetes API server URL could not be constructed from env vars")
+	errInvalidCertificate   = errors.New("invalid CA")
 )
 
 // New creates and initializes a Kubernetes DataClient.
-func New(o Options) *Client {
-	if o.KubernetesURL == "" {
-		o.KubernetesURL = defaultKubernetesURL
+func New(o Options) (*Client, error) {
+	httpClient, err := buildHTTPClient(serviceAccountDir+serviceAccountRootCAKey, o.KubernetesInCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	apiURL, err := buildAPIURL(o)
+	if err != nil {
+		return nil, err
 	}
 
-	log.Debugf("kube client initialized with api address: %s; with healthcheck: %t",
-		o.KubernetesURL, o.ProvideHealthcheck)
+	token, err := readServiceAccountToken(serviceAccountDir+serviceAccountTokenKey, o.KubernetesInCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("running in-cluster: %t. api server url: %s. provide health check: %t", o.KubernetesInCluster, apiURL, o.ProvideHealthcheck)
+
 	return &Client{
-		apiURL:             o.KubernetesURL,
+		httpClient:         httpClient,
+		apiURL:             apiURL,
 		provideHealthcheck: o.ProvideHealthcheck,
 		current:            make(map[string]*eskip.Route),
+		token:              token,
+	}, nil
+}
+
+func readServiceAccountToken(tokenFilePath string, inCluster bool) (string, error) {
+	if !inCluster {
+		return "", nil
+	}
+
+	bToken, err := ioutil.ReadFile(tokenFilePath)
+	if err != nil {
+		return "", err
+	}
+
+	return string(bToken), nil
+}
+
+func buildHTTPClient(certFilePath string, inCluster bool) (*http.Client, error) {
+	if !inCluster {
+		return http.DefaultClient, nil
+	}
+
+	rootCA, err := ioutil.ReadFile(certFilePath)
+	if err != nil {
+		return nil, err
 	}
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(rootCA) {
+		return nil, errInvalidCertificate
+	}
+
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		RootCAs:    certPool,
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	return &http.Client{
+		Transport: transport,
+	}, nil
+}
+
+func buildAPIURL(o Options) (string, error) {
+	if !o.KubernetesInCluster {
+		if o.KubernetesURL == "" {
+			return defaultKubernetesURL, nil
+		}
+		return o.KubernetesURL, nil
+	}
+
+	host, port := os.Getenv(serviceHostEnvVar), os.Getenv(servicePortEnvVar)
```

Comment:
https://github.com/kubernetes/kubernetes/blob/1a46a167f3c1baecb7e0e6a84137330982861ec2/staging/src/k8s.io/client-go/rest/config.go#L278 
the env vars are injected by kubernetes, i just followed the way it is done in the client-go :)